draft-ietf-jsonpath-base.md

```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
 The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
 obey its laws (see, for example, {{BOOLEAN-LAWS}}).
 
+##### Function Extensions {#fnex}
+{: unnumbered}
+
+[^unnumbered-bad]
+
+[^unnumbered-bad]: This should not be an unnumbered section, which it is
+    forced to be by the parent section being unnumbered too.
+    The overall structure where the parts of this subsection go to
+    needs to be decided; one part already is in the IANA
+    Considerations in {{iana-fnex}}
+{:source=" -- cabo"}
+
+Beyond the filter expression functionality defined in the preceding
+subsections, JSONPath defines an extension point that can be used to
+add filter expression functionality: "Function Extensions".
+
+A function extension defines a registered name (see {{iana-fnex}}) that
+can be applied to a sequence of zero or more arguments, resulting in a result.
+A function-expression stands for ("returns") a nodelist or a value.
+The arguments are filter expressions that are to be interpreted as
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
+
+~~~ abnf
+function-name           = function-name-first *function-name-char
+function-name-first     = LCALPHA
+function-name-char      = DIGIT / function-name-first / "_"
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" [function-argument *("," S function-argument)] ")"
+singular-path           =/ function-expression
+function-argument       = comparable
+~~~
+
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable, regex-expr).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable or a
+regex-expr.
+
+###### Function Extensions: length {#length}
+{: unnumbered}
+
+When defining an extension point in a protocol, it is generally a good
+idea to provide at least one extension that makes use of the extension
+point, to serve as an example and to provide motivation to implement
+the extension point.
+
+The "length" function extension provides a way to compute the length
+of an array and make that available for further processing in the
+filter expression:
+
+~~~ JSONPath
+$[?length(@.authors) >= 5]
+~~~
+
+Its only argument is a value (possibly taken from a singular path as
+in the example above).  The result also is a value; if the argument
+value is an array, the result is the number of elements in the array.
```

Comment:
Do we also want to support strings or objects?  They're fairly well-defined for this function.

Reply:
Very good question.  What do implementations out there do?
For strings, I think we'd count the Unicode scalars.
For JSON objects, the number of name-value pairs.

Reply:
We even could allow nodelists and return the number of nodes.

Reply:
That would need to be a separate extension, as length(X) is ambiguous if X can be interpreted as a node that contains an array (which gets us the size of the array) or a nodelist of size 1 (which gets us 1).

Reply:
Separate function (called "count") provided now.